File: #0300.longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {

        int size = nums.size();
        if (size == 0) {
            return 0;
        }

        vector<int> v(size, 1);
        int result = 1;
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    v[i] = std::max(v[i], v[j] + 1);
                }
            }
            result = std::max(result, v[i]);
        }
        return result;
    }
};

File: #0300.longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    let piles = [];
    for (let num of nums) {
        let i = piles.findIndex(pile => pile[0] >= num);
        if (i === -1) {
            piles.push([num]);
        } else {
            piles[i].unshift(num);
        }
    }
    return piles.length;
};
